maml_rl/metalearner.py

```diff
@@ -106,14 +106,14 @@ def kl_divergence(self, episodes, old_pis=None):
 
     def hessian_vector_product(self, episodes, damping=1e-2):
         """Hessian-vector product, based on the Perlmutter method."""
+        kl = self.kl_divergence(episodes)
+        grads = torch.autograd.grad(kl, self.policy.parameters(),
+            create_graph=True)
+        flat_grad_kl = parameters_to_vector(grads)
         def _product(vector):
-            kl = self.kl_divergence(episodes)
-            grads = torch.autograd.grad(kl, self.policy.parameters(),
-                create_graph=True)
-            flat_grad_kl = parameters_to_vector(grads)
-
             grad_kl_v = torch.dot(flat_grad_kl, vector)
-            grad2s = torch.autograd.grad(grad_kl_v, self.policy.parameters())
+            grad2s = torch.autograd.grad(grad_kl_v, self.policy.parameters(),
+                                         retain_graph=True)
```

Comment:
Can you comment on why you add `retain_graph=True` here? I am not very familiar with `retain_graph` so I don't know the proper way to use it, but is it possible to add this to the computation of `grads` instead (where we already create the graph), since that is the only part that would remain fixed?

Reply:
Calling `torch.autograd.grad` without `retain_graph=True` will free the "buffers", which I believe refer to the outputs of each computation in the forward pass. So the `retain_graph=True` needs to be on this call to `torch.autograd.grad`, because otherwise it would free the buffers related to the original forward computations within `kl_divergence` and subsequent calls to `torch.autograd.grad` within `_product` would fail.

The first call to `torch.autograd.grad` that computes the `grads` also has `retain_graph=True` because it defaults to the value of `create_graph`.